flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java

```diff
@@ -0,0 +1,811 @@
+      if (flinkWriteConf.upsertMode()) {
+        Preconditions.checkState(
+            !flinkWriteConf.overwriteMode(),
+            "OVERWRITE mode shouldn't be enable when configuring to use UPSERT data stream.");
+        Preconditions.checkState(
+            !equalityFieldIds.isEmpty(),
+            "Equality field columns shouldn't be empty when configuring to use UPSERT data stream.");
+        if (!table.spec().isUnpartitioned()) {
+          for (PartitionField partitionField : table.spec().fields()) {
+            Preconditions.checkState(
+                equalityFieldIds.contains(partitionField.sourceId()),
+                "In UPSERT mode, partition field '%s' should be included in equality fields: '%s'",
+                partitionField,
+                equalityFieldColumns);
+          }
+        }
+      }
+
+      Map<String, String> writeProperties =
+          writeProperties(table, flinkWriteConf.dataFileFormat(), flinkWriteConf);
+      RowType flinkRowType = toFlinkRowType(table.schema(), tableSchema);
+
+      Duration tableRefreshInterval = flinkWriteConf.tableRefreshInterval();
+      SerializableSupplier<Table> tableSerializableSupplier;
+      if (tableRefreshInterval != null) {
+        tableSerializableSupplier =
+            new CachingTableSupplier(
+                (SerializableTable) table, tableLoader(), tableRefreshInterval);
+      } else {
+        tableSerializableSupplier = new SimpleTableSupplier((SerializableTable) table);
+      }
+
+      Table initTable = tableSerializableSupplier.get();
+      uidPrefix = Optional.ofNullable(uidPrefix).orElse("");
+
+      // FlinkWriteConf properties needed to be set separately, so we do not have to serialize the
+      // full conf
+      boolean upsertMode = flinkWriteConf.upsertMode();
+      FileFormat dataFileFormat = flinkWriteConf.dataFileFormat();
+      long targetDataFileSize = flinkWriteConf.targetDataFileSize();
+      return new IcebergSink(
+          initTable,
+          tableLoader,
+          table,
+          snapshotSummary,
+          uidPrefix,
+          writeProperties,
+          flinkRowType,
+          tableSerializableSupplier,
+          flinkWriteConf,
+          equalityFieldIds,
+          upsertMode,
+          dataFileFormat,
+          targetDataFileSize,
+          branch,
+          overwriteMode,
+          workerPoolSize);
+    }
+
+    /**
+     * Append the iceberg sink operators to write records to iceberg table.
+     *
+     * @return {@link DataStreamSink} for sink.
+     */
+    public DataStreamSink<RowData> append() {
+      IcebergSink sink = build();
+      DataStream<RowData> rowDataInput = inputCreator.apply(uidPrefix);
+      DataStreamSink<RowData> rowDataDataStreamSink =
+          rowDataInput.sinkTo(sink).uid(uidPrefix + "-sink");
+      if (sink.flinkWriteConf.writeParallelism() != null) {
+        rowDataDataStreamSink.setParallelism(sink.flinkWriteConf.writeParallelism());
```

Comment:
This is setting the parallelism of `DataStreamSink` (not the writer tasks). We still need to figure out in v2 sink how is the writer parallelism set.

Reply:
The parallelism of the SinkV2 writer is inherited from the SinkV2 parallelism.
We should check it, but I'm fairly sure reading from the code.
@afedulov could help confirm this, if he is available.

Reply:
we use 
```java
 public Builder writeParallelism(int newWriteParallelism)
```
in the `IcebergSink` Builder

Reply:
I know users can pass in `writeParallelism` via builder. but I am saying it is not applied to the writer operator. it is currently applied to `rowDataDataStreamSink.setParallelism(sink.flinkWriteConf.writeParallelism());` after the `sinkTo`

Reply:
I think this is the relevant part (from the `org.apache.flink.streaming.runtime.translators.SinkTransformationTranslator`):
```
                // remember that the environment parallelism has been set to be default
                // at the beginning. SubTransformations, whose parallelism has been
                // customized, will skip this part. The customized parallelism value set by user
                // will therefore be kept.
                if (subTransformation.getParallelism() == ExecutionConfig.PARALLELISM_DEFAULT) {
                    // In this case, the subTransformation does not contain any customized
                    // parallelism value and will therefore inherit the parallelism value
                    // from the sinkTransformation.
                    subTransformation.setParallelism(transformation.getParallelism());
                }
```

Reply:
I am not sure how to interpret the code snippet. are we saying `rowDataDataStreamSink.setParallelism(sink.flinkWriteConf.writeParallelism());` would propagate the proper parallelism to the writer operator transformation? If yes, let' add a unit test to verify it.

Reply:
I did a debugging where I could assert that the parallelism of the framework operators matched the one set with `IcebergSink.writeParallelism($parallelism)`

Reply:
thanks for confirming. Please add a code comment to explain the behavior. 

Just to clarify, Flink v2 sink framework propagate the parallelism set in this step to other framework operators (like writers, committers). this may need to be revisited if Flink can support setting committer parallelism (to 1 for Iceberg sink). 
```
 rowDataDataStreamSink.setParallelism(sink.flinkWriteConf.writeParallelism());
```